grudge/reductions.py

```diff
@@ -246,10 +257,35 @@ def integral(dcoll: DiscretizationCollection, dd, vec) -> float:
 
 # {{{  Elementwise reductions
 
-def _map_elementwise_reduction(actx: ArrayContext, op_name):
-    @memoize_in(actx, (_map_elementwise_reduction,
+def _apply_elementwise_reduction(
+        op_name: str, dcoll: DiscretizationCollection, *args) -> DOFArray:
+    r"""Returns a vector of DOFs with all entries on each element set
+    to the reduction operation *op_name* over all degrees of freedom.
+
+    :arg \*args: Arguments for the reduction operator, such as *dd* and *vec*.
+    :returns: a :class:`~meshmode.dof_array.DOFArray` or object arrary of
+        :class:`~meshmode.dof_array.DOFArray`s.
+    """
+    if len(args) == 1:
+        vec, = args
+        dd = dof_desc.DOFDesc("vol", dof_desc.DISCR_TAG_BASE)
+    elif len(args) == 2:
+        dd, vec = args
+    else:
+        raise TypeError("invalid number of arguments")
+
+    dd = dof_desc.as_dofdesc(dd)
+
+    if isinstance(vec, np.ndarray):
+        return obj_array_vectorize(
```

Comment:
This should really use containers instead of special-casing `numpy` arrays.

Reply:
@inducer I totally agree. I was planning to address this exact case in a separate PR (to address https://github.com/inducer/grudge/issues/97 specifically), which will include the reductions and all DG operators. Does that sound okay to you?

Reply:
Sure, that's alright by me.